text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,236 @@
+
+A manual decode `Decode`  implementation is required to account for version detection and constructing the appropriate variant.
+
+
+#### New primitive for representing the `CoreIndex` commitments.**
+
+```
+pub enum UMPSignal {
+	OnCore(CoreIndex),
+}
+```
+
+### Cumulus primitives
+
+Add a new version of the `ParachainInherentData` structure which includes an additional `core_index` field.
+```
+pub struct ParachainInherentData {
+	pub validation_data: PersistedValidationData,
+	/// A storage proof of a predefined set of keys from the relay-chain.
+	///
+	/// Specifically this witness contains the data for:
+	///
+	/// - the current slot number at the given relay parent
+	/// - active host configuration as per the relay parent,
+	/// - the relay dispatch queue sizes
+	/// - the list of egress HRMP channels (in the list of recipients form)
+	/// - the metadata for the egress HRMP channels
+	pub relay_chain_state: sp_trie::StorageProof,
+	/// Downward messages in the order they were sent.
+	pub downward_messages: Vec<InboundDownwardMessage>,
+	/// HRMP messages grouped by channels. The messages in the inner vec must be in order they
+	/// were sent. In combination with the rule of no more than one message in a channel per block,
+	/// this means `sent_at` is **strictly** greater than the previous one (if any).
+	pub horizontal_messages: BTreeMap<ParaId, Vec<InboundHrmpMessage>>,
+	/// The core index on which the parachain block must be backed
+	pub core_index: CoreIndex,
+}
+```
+
+### UMP transport
+[CandidateCommitments](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L652) remains unchanged as we will store scale encoded `UMPSignal` messages directly in the parachain UMP queue by outputing them in the [upward_messages](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L682). 
+
+
+The UMP queue layout is adjusted to allow the relay chain to receive both the XCM messages and `UMPSignal` messages. We will introduce a message separator that will be implemented as an empty `Vec<u8>`.
+
+The separator marks the end of the XCM messages and the begging of the `UMPSignal` messages.
+
+Example: 
+```
+[ XCM message1, XCM message2, ..., EMPTY message, UMPSignal::CoreIndex ]
+```
+
+### Parachain block validation
+
+Backers will make use of the core index information to validate the blocks during backing and reject blocks if:
+- the `core_index` in descriptor does not match the one in the `UMPSignal`.
+- the `core_index` in the descriptor does not match the core the backing group is assigned to
+- the `session_index` is equal to the session of the `relay_parent` in the descriptor
+
+If core index (and session index) information is not available (backers got an old candidate receipt), there will be no changes compared to current behaviour.
+
+## Drawbacks
+
+The only drawback is that further additions to the descriptor are limited to the amount of remaining unused space.
```

Comment:
One field we should introduce at some later point, is a proper version. The currently introduced one is version 0.

Reply:
Yes, do you think I should detail a bit more or even include a version field already ?

Reply:
Just a hint that after we should introduce a version field once we successfully sunsetted collators sending actual signatures. The version implemented by this RFC will then be version 0.

Reply:
Added a new `Versioning` section under `Compatibility` that describes how versioning will be done with future changes.